File: java/3_LongestSubstringWithoutRepeatingCharacters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] letters = new int[128];
        Arrays.fill(letters, -1);
        int high = 0, idx = 0;
        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if (letters[c] >= 0) 
                idx = Math.max(letters[c] + 1, idx);
            letters[c] = i;
            high = Math.max(high, i - idx + 1);
        }
        return high;
    }
}

File: swift/3_LongestSubstringWithoutRepeatingCharacters.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        var letters = [Character:Int]()
        var high = 0, idx = 0
        for (i, c) in s.enumerated() {
            if let lastLetterIdx = letters[c] {
                idx = max(lastLetterIdx + 1, idx)
            }
            letters[c] = i
            high = max(high, i - idx + 1)
        }
        return high
    }
}
